src/crab/array_domain.cpp

```diff
@@ -447,6 +450,69 @@ std::ostream& operator<<(std::ostream& o, offset_map_t& m) {
     return o;
 }
 
+// Create a new cell that is a subset of an existing cell.
+void array_domain_t::split_cell(NumAbsDomain& inv, data_kind_t kind, int cell_start_index, unsigned int len) {
+    assert(kind == data_kind_t::svalues || kind == data_kind_t::uvalues);
+
+    // Get the values from the indicated stack range.
+    std::optional<linear_expression_t> svalue = load(inv, data_kind_t::svalues, number_t(cell_start_index), len);
+    std::optional<linear_expression_t> uvalue = load(inv, data_kind_t::uvalues, number_t(cell_start_index), len);
+
+    // Create a new cell for that range.
+    offset_map_t& offset_map = lookup_array_map(kind);
+    cell_t new_cell = offset_map.mk_cell(cell_start_index, len);
+    inv.assign(new_cell.get_scalar(data_kind_t::svalues), svalue);
+    inv.assign(new_cell.get_scalar(data_kind_t::uvalues), uvalue);
+}
+
+// Prepare to havoc bytes in the middle of a cell by potentially splitting the cell if it is numeric,
+// into the part to the left of the havoced portion, and the part to the right of the havoced portion.
+void array_domain_t::split_number_var(NumAbsDomain& inv, data_kind_t kind, const linear_expression_t& i,
+                                      const linear_expression_t& elem_size) {
+    assert(kind == data_kind_t::svalues || kind == data_kind_t::uvalues);
+    offset_map_t& offset_map = lookup_array_map(kind);
+    interval_t ii = inv.eval_interval(i);
+    std::optional<number_t> n = ii.singleton();
+    if (!n) {
+        // We can only split a singleton offset.
+        return;
+    }
+    interval_t i_elem_size = inv.eval_interval(elem_size);
+    std::optional<number_t> n_bytes = i_elem_size.singleton();
+    if (!n_bytes) {
+        // We can only split a singleton size.
+        return;
+    }
+    auto size = static_cast<unsigned int>(*n_bytes);
+    offset_t o((uint64_t)*n);
+
+    std::vector<cell_t> cells = offset_map.get_overlap_cells(o, size);
+    for (cell_t const& c : cells) {
+        interval_t intv = c.to_interval();
+        int cell_start_index = (int)*intv.lb().number();
+        int cell_end_index = (int)*intv.ub().number();
+
+        if (!this->num_bytes.all_num(cell_start_index, cell_end_index + 1) || (cell_end_index + 1 < cell_start_index + sizeof(int64_t))) {
+            // We can only split numeric cells of size 8 or less.
+            continue;
+        }
+
+        std::optional<number_t> old_number = inv.eval_interval(c.get_scalar(kind)).singleton();
+        if (!old_number) {
```

Comment:
`old_number` is not used further. Use `is_singleton()`

Reply:
Done